tests/integration/base.py

```diff
@@ -0,0 +1,327 @@
+    except TimeoutExpired:
+        LOG.error(f"Command: {command_list}, TIMED OUT")
+        LOG.error(f"Return Code: {process_execute.returncode}")
+        process_execute.kill()
+        raise
+
+
+class Base:
+    class IntegBase(TestCase):
+        tempdir: Any
+        directory: str
+
+        def setUp(self) -> None:
+            self.tempdir = tempfile.TemporaryDirectory()
+            self.cwd = Path(self.tempdir.name, PROJECT_NAME)
+            LOG.info(f"Temporary directory {self.tempdir.name} created")
+
+        def tearDown(self) -> None:
+            LOG.info(f"Clean up temporary directory {self.tempdir.name}")
+            self.tempdir.cleanup()
+
+        def _test_init_template(self):
+            template_path = Path(REPO_ROOT, self.directory)
+            cmdlist = [
+                "sam",
+                "init",
+                "--no-input",
+                "--location",
+                template_path,
+                "--name",
+                PROJECT_NAME,
+            ]
+            LOG.info(cmdlist)
+            run_command(cmdlist, self.tempdir.name)
+            self.assertTrue(self.cwd.exists())
+
+    class BuildInvokeBase(IntegBase):
+        function_id_by_event: Optional[Dict[str, str]] = None
+        invoke_output: Dict[str, Any]
+        use_container: bool = True
+
+        def _test_build(self):
+            cmdlist = ["sam", "build", "--debug"]
+            if self.use_container:
+                cmdlist.append("--use-container")
+            LOG.info(cmdlist)
+            result = run_command(cmdlist, self.cwd)
+            self.assertIn("Build Succeeded", str(result.stdout))
+
+        def _test_local_invoke(self):
+            events_path = Path(self.cwd, "events")
+            if not events_path.exists():
+                LOG.info(f"Skip event testing, {events_path} does not exist")
+                return
+            event_files = os.listdir(events_path)
+            for event_file in event_files:
+                if self.function_id_by_event:
+                    cmdlist = [
+                        "sam",
+                        "local",
+                        "invoke",
+                        self.function_id_by_event[event_file],
+                        "-e",
+                        Path("events", event_file),
+                    ]
+                else:
+                    cmdlist = [
+                        "sam",
+                        "local",
+                        "invoke",
+                        "-e",
+                        Path("events", event_file),
+                    ]
+                LOG.info(cmdlist)
+                result = run_command(cmdlist, self.cwd)
+                try:
+                    self.invoke_output = json.loads(result.stdout)
+                except json.decoder.JSONDecodeError:
+                    self.fail(f"Response is not a valid JSON: {result.stdout}")
+
+        @pytest.mark.flaky(reruns=3)
```

Comment:
Curious, why is this marked flaky?

Reply:
I found there are a few reason that cause these build/invoke test to fail randomly.

For example, for .net 2.1, I encountered this:

https://us-west-2.console.aws.amazon.com/codesuite/codebuild/746828864014/projects/CodeBuild-uuRAqDCqptB1/build/CodeBuild-uuRAqDCqptB1%3A63ca3055-e841-4400-972c-328084018a1f/?region=us-west-2

it failed twice due to this error:

<img width="1444" alt="image" src="https://user-images.githubusercontent.com/3804518/101536683-2f146f00-394f-11eb-8622-b235475dbcc1.png">

and in the 3rd run (rerun), it finally passed.

<img width="1654" alt="image" src="https://user-images.githubusercontent.com/3804518/101536816-5cf9b380-394f-11eb-97ca-55e5a20d5cc9.png">


Another example, it takes 1 rerun to pass: https://us-west-2.console.aws.amazon.com/codesuite/codebuild/746828864014/projects/CodeBuild-uuRAqDCqptB1/build/CodeBuild-uuRAqDCqptB1%3Accdb3508-a364-4b5b-aa92-6db3fd58fb81/?region=us-west-2

Another example, .net 3.1 this time, 3 reruns all fail 😢:

https://us-west-2.console.aws.amazon.com/codesuite/codebuild/746828864014/projects/CodeBuild-uuRAqDCqptB1/build/CodeBuild-uuRAqDCqptB1%3A01c29be8-289f-42a4-b836-597000b85632/?region=us-west-2

Reply:
I remember I see other issues related to docker that could also fail randomly. 

In our sam-cli `test_cmd_build`, we also have the same `flaky` decorator, so I copied it here:

<img width="670" alt="image" src="https://user-images.githubusercontent.com/3804518/101537228-f9bc5100-394f-11eb-81f9-be8c8923c5db.png">